crates/catalog/sql/src/catalog.rs

```diff
@@ -167,43 +177,335 @@ impl SqlCatalog {
             .await
             .map_err(from_sqlx_error)
     }
+
+    /// Execute statements in a transaction, provided or not
+    pub async fn execute(
+        &self,
+        query: &str,
+        args: Vec<Option<&str>>,
+        transaction: Option<&mut Transaction<'_, Any>>,
+    ) -> Result<AnyQueryResult> {
+        let query_with_placeholders = self.replace_placeholders(query);
+
+        let mut sqlx_query = sqlx::query(&query_with_placeholders);
+        for arg in args {
+            sqlx_query = sqlx_query.bind(arg);
+        }
+
+        match transaction {
+            Some(t) => sqlx_query.execute(&mut **t).await.map_err(from_sqlx_error),
+            None => {
+                let mut tx = self.connection.begin().await.map_err(from_sqlx_error)?;
+                let result = sqlx_query.execute(&mut *tx).await.map_err(from_sqlx_error);
+                let _ = tx.commit().await.map_err(from_sqlx_error);
+                result
+            }
+        }
+    }
 }
 
 #[async_trait]
 impl Catalog for SqlCatalog {
     async fn list_namespaces(
         &self,
-        _parent: Option<&NamespaceIdent>,
+        parent: Option<&NamespaceIdent>,
     ) -> Result<Vec<NamespaceIdent>> {
-        todo!()
+        let table_namespaces_stmt = format!(
+            "SELECT {CATALOG_FIELD_TABLE_NAMESPACE}
```

Comment:
The DISTINCT is left out on purpose as the UNION later on removes duplicates across both of the queries. I'll see if this is easier to do client-side when I make the whole function easier to read.

Reply:
I've had another look through, I could add some more comments to clear this up that the UNION will remove duplicates. I think it's far easier to have the database do the uniqueness check as opposed to code to either run 2 queries, then union on the client, or force the query to return all rows (UNION ALL) then do the unique check on the client.

Reply:
Ok nevermind, I have changed this - we can do the parent filtering with the query return instead of pushing it to SQL. See https://github.com/apache/iceberg-rust/pull/534/commits/b0b731c2ed2cab8ffd3df381b17f989e8b8f6380